packages/gatsby/src/query/index.js

```diff
@@ -305,7 +307,10 @@ const startListeningToDevelopQueue = ({ graphqlTracing } = {}) => {
   listenerQueue = new Queue((queryJobs, callback) => {
     const activity = createQueryRunningActivity(queryJobs.length)
 
-    const onFinish = (...arg) => {
+    const onFinish = async (...arg) => {
+      if (!webpackStatusUtil.isPending()) {
+        await pageDataUtil.flush()
```

Comment:
That doesn't set status to pending, but calls flush.

Reply:
Because queries can be run without a webpack run

Reply:
Right and we can't block webpack anyway. However wouldn't that potentially cause a race condition if webpack thing happens in the middel of this flush and calls its own flush?

Reply:
Oh yeah, you're right. We need a `currently flushing` state so that we don't run more than 1 at once, I think. 

@pieh Thoughts?

Reply:
I think we need a (for now singleton) Flusher object for which "enqueueFlush" can be called. When webpackStatusUtil is not pending and it's not flushing, it will flush at once. Otherwise at the end of flush it will check if it can flush again. Because Flusher needs to know about webpackStatusUtil, I assume they should be connected or be same object.

Reply:
Need to think about this a bit - we might need to reach to `suspend` / `resume` earlier than expected (I will need to anyway, so this might be great time to explore this together)

Reply:
so smth like that

```
Flusher.enqueueFlush()
Flusher.setWebpackPending()
Flusher.setWebpackDone()

setWebpackDone() {
  this.webpackStatus = `done`
  if (this.enqueuedFlush) {
    this.flushing = true
    await pageDataUtil.flush()
    this.flushing = false
  }
}
```

Reply:
@freiksenet Indeed. Thanks for catching this! I'll test and clean this up later today/Monday

Reply:
I think with this we can avoid suspend/resume.

Reply:
Yeah, we shouldn't need to suspend/resume. The point is we were checking webpack running state but not flushing running state (which is also async) 

pseudocode from Mikhail will fix, I'll update the PR

Reply:
I see what you mean. NIT: but pseudocode never actually check for `this.flushing` anywhere - so need to add this somewhere



(and doesn't match description of previous comment :P)

Reply:
Fixed